### PR TITLE
Sbee persist gain

### LINKED ISSUE #173: Persist gain_db in enlighten.ini

### PR DESCRIPTION
**Pull request by samiebee43**
Close #173

### CODE REVIEW

File: enlighten/scope/GainDBFeature.py [hunk 1]
```diff
@@ -121,31 +123,39 @@ def reset(self, hotplug=False):
...
+        self.multispec.set_state("gain_db", db)
+        self.multispec.change_device_setting("detector_gain", db)
+
+        # ensure both gain widgets are correct, without generating additional events
+        self._quiet_set(self.spinbox, db)
+        self._quiet_set(self.slider, db)
+
+        # save gain_db to .ini
+        log.debug("Save gain to INI: %s", db)
+        self.ctl.config.set("detector", "gain_db", float(db))
```
Comment (mzieg) [inline enlighten/scope/GainDBFeature.py:155]: The problem is, this would force every SiG to use the same gain.  Normally we store and load spectrometer settings on a per-serial number basis.

Settings are normally saved in Configuration.save_file, and loaded (this could arguably move to Spectrometer) in Controller.set_from_ini_file().

**Review by mzieg — CHANGES_REQUESTED** [review]

File: enlighten/scope/GainDBFeature.py (hunk 1 above)
Comment (samiebee43) [reply enlighten/scope/GainDBFeature.py:155]: > Normally we store and load spectrometer settings on a per-serial number basis.

Does this apply to #156, #158, #164 ?

Comment (mzieg) [reply enlighten/scope/GainDBFeature.py]: Good question.  I would say no, and here's my thinking:

- integration time and gain are very much "device settings" — they're set within the spectrometer, they have "startup" fields in the EEPROM to initialize their per-device settings, etc.  Any setting within ENLIGHTEN which more-or-less directly maps to a hardware setting or existing EEPROM field should probably be tracked per-device.

- plugins (#158) and the cursor (#156 and #164) feel, to me, more like "ENLIGHTEN settings." I think it's fair to treat those as application-level settings.  That said, this could be debated.

To me, the .ini format itself somewhat leans toward "simple settings" (integration time, gain, anything which needs to store exactly one scalar) being "convenient" for per-serial number store, while "complicated settings" where you have to store more than one value, and maybe include sets of values are simpler if given their own Configuration section.

Basically, I'm ambivalent on this.  I personally think it would be easier and more intuitive if Cursor and Plugins were application-wide, but if someone else had a strong opinion otherwise I could be persuaded.

**Comment by samiebee43** [comment]
Need a second spec to test this.

Specifically, looking at multispec load from .ini -- does GainDBFeature.reset get called every time user changes current_spec?

**Review by mzieg — APPROVED** [review]

**Comment by samiebee43** [comment]
## Acceptance Criteria

- (apply debug to show gain for FX2)
- Open Enlighten
- Set Gain to 32
- Close Enlighten
- Open it again
- Check that Gain is 32

This is a test after the change to be specific to serial numbers. I did not test this with multiple spectrometers connected.

File: enlighten/scope/GainDBFeature.py (hunk 1 above)
Comment (mzieg) [reply enlighten/scope/GainDBFeature.py]: We should probably also persist Scan Averaging, and your question remains pertinent: should that be per-serial number or "application-wide?"  In this case I'm going to go with per-serial number, because on at least some spectrometers (SiG), that _will_ be an attribute written to the firmware via USB/BLE.  (And currently it's handled within WasatchDevice, again a per-unit association.)
